File: examples/go/minimal-order-service/application/usecases/place_order.go
package usecases

import (
	"fmt"
	"github.com/example/order-service/application/ports/inbound"
	"github.com/example/order-service/application/ports/outbound"
	"github.com/example/order-service/domain"
)

// PlaceOrderUseCase implements the PlaceOrderPort.
type PlaceOrderUseCase struct {
	orderRepository outbound.OrderRepository
	paymentGateway  outbound.PaymentGateway
}

// NewPlaceOrderUseCase creates a new PlaceOrderUseCase.
func NewPlaceOrderUseCase(
	orderRepository outbound.OrderRepository,
	paymentGateway outbound.PaymentGateway,
) *PlaceOrderUseCase {
	return &PlaceOrderUseCase{
		orderRepository: orderRepository,
		paymentGateway:  paymentGateway,
	}
}

// PlaceOrder executes the place order use case.
func (uc *PlaceOrderUseCase) PlaceOrder(command inbound.PlaceOrderCommand) (*domain.Order, error) {
	// Create domain object
	total, err := domain.NewMoney(command.Amount, command.Currency)
	if err != nil {
		return nil, fmt.Errorf("invalid money: %w", err)
	}

	order, err := domain.CreateOrder(total)
	if err != nil {
		return nil, fmt.Errorf("failed to create order: %w", err)
	}

	// Process payment through port
	paymentResult, err := uc.paymentGateway.ProcessPayment(total, order.ID())
	if err != nil {
		return nil, fmt.Errorf("payment processing failed: %w", err)
	}

	if !paymentResult.Success {
		return nil, fmt.Errorf("payment failed: %s", paymentResult.ErrorMessage)
	}

	// Mark order as paid
	if err := order.MarkAsPaid(); err != nil {
		return nil, fmt.Errorf("failed to mark order as paid: %w", err)
	}

	// Save through port
	if err := uc.orderRepository.Save(order); err != nil {
		return nil, fmt.Errorf("failed to save order: %w", err)
	}

	return order, nil
}

File: examples/python/minimal-order-service/application/usecases/place_order.py
"""
Use case: Place Order
Orchestrates domain objects and outbound ports to fulfill the business requirement.
"""
from domain.money import Money
from domain.order import Order
from application.ports.inbound.place_order_port import PlaceOrderPort, PlaceOrderCommand
from application.ports.outbound.order_repository import OrderRepository
from application.ports.outbound.payment_gateway import PaymentGateway


class PlaceOrderUseCase(PlaceOrderPort):
    """Use case: Place Order"""
    
    def __init__(
        self,
        order_repository: OrderRepository,
        payment_gateway: PaymentGateway
    ):
        """
        Initialize the use case.
        
        Args:
            order_repository: Order repository port
            payment_gateway: Payment gateway port
        """
        self._order_repository = order_repository
        self._payment_gateway = payment_gateway
    
    def place_order(self, command: PlaceOrderCommand) -> Order:
        """
        Execute the place order use case.
        
        Args:
            command: Place order command
        
        Returns:
            Created order
        
        Raises:
            ValueError: If payment fails
        """
        # Create domain object
        total = Money(command.amount, command.currency)
        order = Order.create(total)
        
        # Process payment through port
        payment_result = self._payment_gateway.process_payment(total, order.id)
        
        if not payment_result.success:
            raise ValueError(f"Payment failed: {payment_result.error_message}")
        
        # Mark order as paid
        order.mark_as_paid()
        
        # Save through port
        self._order_repository.save(order)
        
        return order
